source/dofs/dof_tools_constraints.cc

```diff
@@ -2442,18 +2442,30 @@ namespace DoFTools
     const ComponentMask &            component_mask,
     const std::vector<unsigned int> &first_vector_components)
   {
-    using FaceVector = std::vector<
-      GridTools::PeriodicFacePair<typename DoFHandlerType::cell_iterator>>;
-    typename FaceVector::const_iterator it, end_periodic;
-    it           = periodic_faces.begin();
-    end_periodic = periodic_faces.end();
+    //
+    // Make sure the user does not accidentally run into a nasty bug by
+    // calling make_hanging_node_constraints first. See issue #7053.
+    //
+    Assert(
+      constraints.n_constraints() == 0,
+      ExcMessage(
+        "Called with a nonempty AffineConstraints object. "
+        "\n\n"
+        "This check is here to guard against a nasty bug that occurs in "
+        "parallel computations when make_hanging_node_constraints is "
+        "accidentally called prior to make_periodicity_constraints. This "
+        "does not work, you have to call make_periodicity_constraints first. "
+        "If you really need to call this function with a non-empty "
+        "AffineConstraints object, create an empty temporary "
+        "AffineConstraints object first, call this function, and then "
+        "merge it back into the original AffineConstraints container."));
```

Comment:
Hmm.... This will break codes. I am in favor of merging back automatically if possible. Do you expect that to give correct results as well?

Reply:
I guess what we can do is to send an empty object to store periodic constraints and merge the user-provided one once constraints are collected. But either way, this is most likely an incompatible change, which may or may not give the same results as users previously had. From that perspective, I am fine to through an error.

Reply:
But there is a difference in returning different constraints and throwing an error and aborting the program. If we can start with an empty object and then merge, I would really prefer that. I can give it a try.

Reply:
This assert cannot work - we have to support for example to call the `make_periodicity_constraints` functions multiple times. But yes, we could simply merge in this very function.